File: server/whatsapp.js
import { db } from './db.js';

/** Normaliza teléfonos AR a formato internacional sin + (54911...) */
export function normalizePhone(raw) {
  if (!raw) return null;
  let digits = String(raw).replace(/\D/g, '');
  if (!digits) return null;
  if (digits.startsWith('00')) digits = digits.slice(2);
  if (digits.startsWith('54')) {
    // ok
  } else if (digits.startsWith('0')) {
    digits = `54${digits.slice(1)}`;
  } else if (digits.length <= 10) {
    digits = `54${digits}`;
  }
  // WhatsApp AR móvil suele ir con 9 después del 54
  if (digits.startsWith('54') && !digits.startsWith('549') && digits.length >= 12) {
    digits = `549${digits.slice(2)}`;
  }
  return digits;
}

export function telHref(raw) {
  const digits = String(raw || '').replace(/\D/g, '');
  if (!digits) return null;
  return `tel:${digits}`;
}

function fillTemplate(body, vars) {
  return String(body || '').replace(/\{(\w+)\}/g, (_, key) => {
    const v = vars[key];
    return v == null || v === '' ? '' : String(v);
  });
}

export function orderVars(order) {
  return {
    cliente: order.client_name || '',
    trabajo: order.title || '',
    estado: order.status_name || '',
    fecha: order.scheduled_date || '',
    hora: order.scheduled_time || '',
    tecnico: order.technician_name || '',
    direccion: [order.client_address, order.locality].filter(Boolean).join(', '),
    telefono: order.client_phone || '',
    orden: String(order.id || ''),
  };
}

export function buildWhatsAppUrl(phone, text) {
  const n = normalizePhone(phone);
  if (!n) return null;
  return `https://wa.me/${n}?text=${encodeURIComponent(text || '')}`;
}

export function listWhatsAppTemplates() {
  return db.prepare(
    `SELECT t.*, s.name AS status_name
     FROM whatsapp_templates t
     LEFT JOIN service_statuses s ON s.id = t.status_id
     ORDER BY t.id`
  ).all();
}

export function messageForOrder(order, templateId = null) {
  let tpl = null;
  if (templateId) {
    tpl = db.prepare('SELECT * FROM whatsapp_templates WHERE id = ? AND active = 1').get(templateId);
  } else if (order.status_id) {
    tpl = db.prepare(
      `SELECT * FROM whatsapp_templates
       WHERE active = 1 AND status_id = ?
       ORDER BY id LIMIT 1`
    ).get(order.status_id);
  }
  if (!tpl) {
    tpl = db.prepare(
      `SELECT * FROM whatsapp_templates WHERE active = 1 AND status_id IS NULL ORDER BY id LIMIT 1`
    ).get();
  }
  if (!tpl) return null;
  const text = fillTemplate(tpl.body, orderVars(order));
  const url = buildWhatsAppUrl(order.client_phone, text);
  if (!url) return { error: 'El cliente no tiene teléfono válido', template: tpl.name, text };
  return { url, text, template: tpl.name, template_id: tpl.id, auto_open: !!tpl.auto_open };
}

File: backend/src/utils/whatsapp.go
package utils

import (
	"fmt"
	"net/url"
	"regexp"
	"strings"
)

var nonDigit = regexp.MustCompile(`\D+`)

// NormalizePhone converts AR phone numbers to international digits without +.
func NormalizePhone(raw string) string {
	if raw == "" {
		return ""
	}
	digits := nonDigit.ReplaceAllString(raw, "")
	if digits == "" {
		return ""
	}
	if strings.HasPrefix(digits, "00") {
		digits = digits[2:]
	}
	if !strings.HasPrefix(digits, "54") {
		if strings.HasPrefix(digits, "0") {
			digits = "54" + digits[1:]
		} else if len(digits) <= 10 {
			digits = "54" + digits
		}
	}
	if strings.HasPrefix(digits, "54") && !strings.HasPrefix(digits, "549") && len(digits) >= 12 {
		digits = "549" + digits[2:]
	}
	return digits
}

func TelHref(raw string) *string {
	digits := nonDigit.ReplaceAllString(raw, "")
	if digits == "" {
		return nil
	}
	s := "tel:" + digits
	return &s
}

func FillTemplate(body string, vars map[string]string) string {
	re := regexp.MustCompile(`\{(\w+)\}`)
	return re.ReplaceAllStringFunc(body, func(m string) string {
		key := m[1 : len(m)-1]
		if v, ok := vars[key]; ok {
			return v
		}
		return ""
	})
}

func OrderWhatsAppVars(order map[string]any) map[string]string {
	str := func(key string) string {
		if v, ok := order[key]; ok && v != nil {
			return fmt.Sprint(v)
		}
		return ""
	}
	addr := strings.TrimSpace(strings.Join(
		filterNonEmpty(str("client_address"), str("locality")),
		", ",
	))
	return map[string]string{
		"cliente":   str("client_name"),
		"trabajo":   str("title"),
		"estado":    str("status_name"),
		"fecha":     str("scheduled_date"),
		"hora":      str("scheduled_time"),
		"tecnico":   str("technician_name"),
		"direccion": addr,
		"telefono":  str("client_phone"),
		"orden":     str("id"),
	}
}

func filterNonEmpty(parts ...string) []string {
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if strings.TrimSpace(p) != "" && p != "<nil>" {
			out = append(out, p)
		}
	}
	return out
}

func BuildWhatsAppURL(phone, text string) string {
	n := NormalizePhone(phone)
	if n == "" {
		return ""
	}
	return "https://wa.me/" + n + "?text=" + url.QueryEscape(text)
}
